flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamTaskStateInitializerImpl.java

```diff
@@ -258,6 +258,10 @@ protected OperatorStateBackend operatorStateBackend(
 		}
 	}
 
+	protected TtlTimeProvider getTtlTimeProvider() {
```

Comment:
Why not to inject this in the constructor of `StreamTaskStateInitializerImpl`?

Reply:
I just don't want to introduce too many changes here. If we plan to inject this as a filed, I am in favor of moving `TtlTimeProvider` at operator level (by changing `StreamTaskStateInitializer#streamOperatorStateContext`) but not in the constructor which was [FLINK-14156](https://issues.apache.org/jira/browse/FLINK-14156) fixed.

Reply:
I think either approach can be refactored in future.
Using constructor is the usual way how we inject dependencies for tests. The inheritance is usually used to change more complicated internal behaviour.
You could also keep the current constructor and add another one annotated with @VisibleForTesting which would accept the custom `TtlTimeProvider`.

Reply:
There existed another problem why we not change the constructor of `StreamTaskStateInitializerImpl`.
Current `AbstractStreamOperatorTestHarness` is not created from a builder, and once a new `AbstractStreamOperatorTestHarness` is created, the inner `streamTaskStateInitializer` has been created with the default `TtlTimeProvider`. Even we set ttl time provider to `AbstractStreamOperatorTestHarness` later, the inner `streamTaskStateInitializer` would not notice the changed ttl time provider unless we call `AbstractStreamOperatorTestHarness#setup` to re-create the inner `streamTaskStateInitializer`. 
However, `AbstractStreamOperatorTestHarness#setup` actually call a deprecated `SetupableStreamOperator#setup` interface. 
In a nutshell, unless we refactor how we build `AbstractStreamOperatorTestHarness`, to make the customized ttl time provider take effect, we must call `AbstractStreamOperatorTestHarness#setup` each time which might already be treated as a deprecated interface.

Reply:
Then I think it makes sense to consider builders to avoid adding more harness constructors.

Reply:
Why not `AbstractStreamOperatorTestHarness` always use a mocked `TtlTimeProvider`?  Just like always use the `TestProcessingTimeService`. Then we don't need to re-create the inner `streamTaskStateInitializer`.

I also prefer the way to set processing time on the harness object `AbstractStreamOperatorTestHarness#setProcessingTime` instead of on other object `mockTtlTimeProvider.setCurrentTimeStamp(0L)`. We can provide a method `setStateTtlTime()` on the `AbstractStreamOperatorTestHarness` too.

Reply:
This would indeed work. This is not flexible in general but may be good enough as we already have it for processing time.

Reply:
I think @wuchong 's suggestion is also a available choice. However, I think FLINK-17011 is the most clean solution to clean up these test code. Although I have to admit that PR https://github.com/apache/flink/pull/11676 is a bit large for reviewing.

Reply:
What about going with this approach first? This can be merged faster and unblock other works. We can also continue the cleanup work using new builders, and expose custom time provider in builders to make it more flexible.